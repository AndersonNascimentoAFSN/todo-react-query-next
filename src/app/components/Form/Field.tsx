import { HTMLAttributes } from "react";

interface FieldProps extends HTMLAttributes<HTMLDivElement> { }

export function Field(props: FieldProps) {
  return (
    <div {...props} className="flex flex-1 flex-col gap-1" />
  )
}