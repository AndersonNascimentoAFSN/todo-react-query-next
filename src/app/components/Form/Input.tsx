import { InputHTMLAttributes, useMemo } from "react";
import { useFormContext } from "react-hook-form";

type InputProps = InputHTMLAttributes<HTMLInputElement> & {
  isError?: boolean
  name: string
};

const getIsErrorClasses = (isError: boolean) => {
  if (isError) {
    return 'border-2 border-red focus:border-red';
  }
}

const BASE_INPUT_CLASSES = 'w-full p-4 rounded-md bg-gray1 placeholder:text-gray5 text-gray6 outline-none focus:border-2 focus:border-green-light';

export const Input = (props: InputProps) => {
  const { isError = false, name = '', ...rest } = props;

  const { register } = useFormContext()

  const computedClasses = useMemo(() => {
    const isErrorClass = getIsErrorClasses(isError);

    return [isErrorClass].join(' ');
  }, [isError]);

  return (
    <input
      {...rest}
      autoComplete="off"
      className={`${BASE_INPUT_CLASSES} ${computedClasses} ${rest?.className}`}
      {...register(name)}
    />
  )
}

Input.displayName = 'Form.Input'
