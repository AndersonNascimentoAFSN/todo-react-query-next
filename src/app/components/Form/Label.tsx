import { LabelHTMLAttributes } from "react";

const BASE_LABEL_CLASSES = '';

export function Label(props: LabelHTMLAttributes<HTMLLabelElement>) {
  return (
    <label
      {...props}
      className={`${BASE_LABEL_CLASSES} ${props?.className}`}
    />
  )
}