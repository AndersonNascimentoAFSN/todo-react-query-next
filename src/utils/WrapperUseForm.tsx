import { FormProvider, useForm } from "react-hook-form";

export function WrapperUseForm(props: any) {
  const formMethods = useForm();

  return (
    <FormProvider {...formMethods}>
      {props.children}
    </FormProvider>
  );
};