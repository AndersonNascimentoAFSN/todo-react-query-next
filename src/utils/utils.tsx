import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import * as React from "react";

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

export function renderWithClient(ui: React.ReactNode) {
  const testQueryClient = createTestQueryClient();
  const { rerender, ...result } = render(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
  );
  return {
    ...result,
    rerender: (rerenderUi: React.ReactNode) =>
      rerender(
        <QueryClientProvider client={testQueryClient}>
          {rerenderUi}
        </QueryClientProvider>
      ),
  };
}

export function createWrapper() {
  const testQueryClient = createTestQueryClient();

  // eslint-disable-next-line react/display-name
  return ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={testQueryClient}>
      {children}
    </QueryClientProvider>
  );
}

createWrapper.displayName = "createWrapper";