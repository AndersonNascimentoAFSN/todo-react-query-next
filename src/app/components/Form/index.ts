import { Input } from './Input';
import { Label } from './Label';
import { Field } from './Field';
import { ErrorMessage } from './ErrorMessage';

export const Form = {
  Input,
  Label,
  Field,
  ErrorMessage,
}