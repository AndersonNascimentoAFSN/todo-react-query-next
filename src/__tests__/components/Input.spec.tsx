import { render, screen, logRoles, act, waitFor } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import { Input } from '@/app/components/Form/Input'
import { WrapperUseForm } from "@/utils/WrapperUseForm"

describe('<Input />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should be able to render the input correctly', () => {
    const { container } = render(
      <WrapperUseForm>
        <Input name="name" aria-label="name" />
      </WrapperUseForm>
    )

    const inputElement = screen.getByRole('textbox', { name: /name/i })

    expect(inputElement).toBeVisible()
    expect(container).toMatchSnapshot()
  })

  it('should be able to write in the input', async () => {
    render(
      <WrapperUseForm>
        <Input name="name" aria-label="name" />
      </WrapperUseForm>
    )

    const inputValue = 'John Doe'

    const inputElement = screen.getByRole('textbox', { name: /name/i })

    await act(async () => {
      await userEvent.type(inputElement, inputValue)
    })

    expect(inputElement).toHaveValue(inputValue)
  })
})