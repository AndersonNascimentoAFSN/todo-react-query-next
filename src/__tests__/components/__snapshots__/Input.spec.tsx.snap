// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should be able to render the input correctly 1`] = `
<div>
  <input
    aria-label="name"
    autocomplete="off"
    class="w-full p-4 rounded-md bg-gray1 placeholder:text-gray5 text-gray6 outline-none focus:border-2 focus:border-green-light  undefined"
    name="name"
  />
</div>
`;
